ophyd/v2/core.py

```diff
@@ -403,86 +408,156 @@ def set_callback(self, callback: Optional[ReadingValueCallback[T]]) -> None:
 }
 
 
+class SimConverter(Generic[T]):
+    def value(self, value: T) -> T:
+        return value
+
+    def write_value(self, value: T) -> T:
+        return value
+
+    def reading(self, value: T, timestamp: float, severity: int) -> Reading:
+        return Reading(
+            value=value,
+            timestamp=timestamp,
+            alarm_severity=-1 if severity > 2 else severity,
+        )
+
+    def descriptor(self, source: str, value) -> Descriptor:
+        assert (
+            type(value) in primitive_dtypes
+        ), f"invalid converter for value of type {type(value)}"
+        dtype = primitive_dtypes[type(value)]
+        return dict(source=source, dtype=dtype, shape=[])
+
+    def make_initial_value(self, datatype: Optional[Type[T]]) -> T:
+        if datatype is None:
+            return cast(T, None)
+
+        return datatype()
+
+
+class SimArrayConverter(SimConverter):
+    def descriptor(self, source: str, value) -> Descriptor:
+        return dict(source=source, dtype="array", shape=[len(value)])
+
+    def make_initial_value(self, datatype: Optional[Type[T]]) -> T:
+        if datatype is None:
+            return cast(T, None)
+
+        if get_origin(datatype) == abc.Sequence:
+            return cast(T, [])
+
+        return cast(T, datatype(shape=0))  # type: ignore
+
+
+@dataclass
+class SimEnumConverter(SimConverter):
+    enum_class: Type[Enum]
+
+    def write_value(self, value: Union[Enum, str]) -> Enum:
+        if isinstance(value, Enum):
+            return value
+        else:
+            return self.enum_class(value)
+
+    def descriptor(self, source: str, value) -> Descriptor:
+        choices = [e.value for e in self.enum_class]
+        return dict(source=source, dtype="string", shape=[], choices=choices)  # type: ignore
+
+    def make_initial_value(self, datatype: Optional[Type[T]]) -> T:
+        if datatype is None:
+            return cast(T, None)
+
+        return cast(T, list(datatype.__members__.values())[0])  # type: ignore
+
+
+class DisconnectedSimConverter(SimConverter):
+    def __getattribute__(self, __name: str) -> Any:
+        raise NotImplementedError("No PV has been set as connect() has not been called")
+
+
+def make_converter(datatype):
+    is_array = get_dtype(datatype) is not None
+    is_sequence = get_origin(datatype) == abc.Sequence
+    is_enum = issubclass(datatype, Enum) if inspect.isclass(datatype) else False
+
+    if is_array or is_sequence:
+        return SimArrayConverter()
+    if is_enum:
+        return SimEnumConverter(datatype)
+
+    return SimConverter()
+
+
 class SimSignalBackend(SignalBackend[T]):
     """An simulated backend to a Signal, created with ``Signal.connect(sim=True)``"""
 
-    _initial_value: T
     _value: T
+    _initial_value: T
     _timestamp: float
+    _severity: int
 
     def __init__(self, datatype: Optional[Type[T]], source: str) -> None:
+        pv = re.split(r"://", source)[-1]
```

Comment:
I'm curious, did you have a reason to use `re.split` rather than `str.split`? Either is fine by me.

Reply:
hm... not particularly. No idea why that's re.split!